File: game-engine/src/test/java/org/luxons/sevenwonders/game/cards/HandsTest.java
package org.luxons.sevenwonders.game.cards;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.junit.Test;
import org.junit.experimental.theories.DataPoints;
import org.junit.experimental.theories.FromDataPoints;
import org.junit.experimental.theories.Theories;
import org.junit.experimental.theories.Theory;
import org.junit.runner.RunWith;
import org.luxons.sevenwonders.game.api.HandCard;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.cards.Hands.PlayerIndexOutOfBoundsException;
import org.luxons.sevenwonders.game.test.TestUtilsKt;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.junit.Assume.assumeTrue;

@RunWith(Theories.class)
public class HandsTest {

    @DataPoints("nbCardsPerPlayer")
    public static int[] nbCardsPerPlayer() {
        return new int[] {0, 1, 2, 3, 4, 5, 6, 7};
    }

    @DataPoints("nbPlayers")
    public static int[] nbPlayers() {
        return new int[] {3, 4, 5, 6, 7};
    }

    private static Hands createHands(int nbPlayers, int nbCardsPerPlayer) {
        Map<Integer, List<Card>> hands = new HashMap<>();
        for (int p = 0; p < nbPlayers; p++) {
            int firstCardNumber = (p - 1) * nbCardsPerPlayer;
            hands.put(p, TestUtilsKt.createSampleCards(firstCardNumber, nbCardsPerPlayer));
        }
        return new Hands(hands, nbPlayers);
    }

    @Test(expected = PlayerIndexOutOfBoundsException.class)
    public void get_failsOnMissingPlayer() {
        Hands hands = createHands(4, 7);
        hands.get(5);
    }

    @Test
    public void get_retrievesCorrectCards() {
        List<Card> hand0 = TestUtilsKt.createSampleCards(0, 5);
        List<Card> hand1 = TestUtilsKt.createSampleCards(5, 10);
        Map<Integer, List<Card>> handsMap = new HashMap<>();
        handsMap.put(0, hand0);
        handsMap.put(1, hand1);
        Hands hands = new Hands(handsMap, 2);
        assertEquals(hand0, hands.get(0));
        assertEquals(hand1, hands.get(1));
    }

    @Theory
    public void isEmpty_falseWhenAtLeast1_allSame(@FromDataPoints("nbPlayers") int nbPlayers,
            @FromDataPoints("nbCardsPerPlayer") int nbCardsPerPlayer) {
        assumeTrue(nbCardsPerPlayer >= 1);
        Hands hands = createHands(nbPlayers, nbCardsPerPlayer);
        assertFalse(hands.isEmpty());
    }

    @Theory
    public void isEmpty_trueWhenAllEmpty(@FromDataPoints("nbPlayers") int nbPlayers) {
        Hands hands = createHands(nbPlayers, 0);
        assertTrue(hands.isEmpty());
    }

    @Theory
    public void maxOneCardRemains_falseWhenAtLeast2_allSame(@FromDataPoints("nbPlayers") int nbPlayers,
            @FromDataPoints("nbCardsPerPlayer") int nbCardsPerPlayer) {
        assumeTrue(nbCardsPerPlayer >= 2);
        Hands hands = createHands(nbPlayers, nbCardsPerPlayer);
        assertFalse(hands.maxOneCardRemains());
    }

    @Theory
    public void maxOneCardRemains_trueWhenAtMost1_allSame(@FromDataPoints("nbPlayers") int nbPlayers,
            @FromDataPoints("nbCardsPerPlayer") int nbCardsPerPlayer) {
        assumeTrue(nbCardsPerPlayer <= 1);
        Hands hands = createHands(nbPlayers, nbCardsPerPlayer);
        assertTrue(hands.maxOneCardRemains());
    }

    @Theory
    public void maxOneCardRemains_trueWhenAtMost1_someZero(@FromDataPoints("nbPlayers") int nbPlayers) {
        Hands hands = createHands(nbPlayers, 1);
        hands.get(0).remove(0);
        assertTrue(hands.maxOneCardRemains());
    }

    @Theory
    public void gatherAndClear(@FromDataPoints("nbPlayers") int nbPlayers,
            @FromDataPoints("nbCardsPerPlayer") int nbCardsPerPlayer) {
        Hands hands = createHands(nbPlayers, nbCardsPerPlayer);
        List<Card> remainingCards = hands.gatherAndClear();
        assertEquals(nbPlayers * nbCardsPerPlayer, remainingCards.size());
        assertTrue(hands.isEmpty());
    }

    @Test
    public void rotate_movesOfCorrectOffset_right() {
        Hands hands = createHands(3, 7);
        Hands rotated = hands.rotate(HandRotationDirection.RIGHT);
        assertEquals(rotated.get(1), hands.get(0));
        assertEquals(rotated.get(2), hands.get(1));
        assertEquals(rotated.get(0), hands.get(2));
    }

    @Test
    public void rotate_movesOfCorrectOffset_left() {
        Hands hands = createHands(3, 7);
        Hands rotated = hands.rotate(HandRotationDirection.LEFT);
        assertEquals(rotated.get(2), hands.get(0));
        assertEquals(rotated.get(0), hands.get(1));
        assertEquals(rotated.get(1), hands.get(2));
    }

    @Test
    public void createHand_containsAllCards() {
        List<Card> hand0 = TestUtilsKt.createSampleCards(0, 5);
        List<Card> hand1 = TestUtilsKt.createSampleCards(5, 10);
        Map<Integer, List<Card>> handsMap = new HashMap<>();
        handsMap.put(0, hand0);
        handsMap.put(1, hand1);
        Hands hands = new Hands(handsMap, 2);

        Table table = TestUtilsKt.testTable(2);
        List<HandCard> hand = hands.createHand(table, 0);

        for (HandCard handCard : hand) {
            assertTrue(hand0.contains(handCard.getCard()));
        }
    }
}

File: game-engine/src/test/kotlin/org/luxons/sevenwonders/game/cards/HandsTest.kt
package org.luxons.sevenwonders.game.cards

import org.junit.Assert.*
import org.junit.Assume.assumeTrue
import org.junit.Test
import org.junit.experimental.theories.DataPoints
import org.junit.experimental.theories.FromDataPoints
import org.junit.experimental.theories.Theories
import org.junit.experimental.theories.Theory
import org.junit.runner.RunWith
import org.luxons.sevenwonders.game.SimplePlayer
import org.luxons.sevenwonders.game.test.sampleCards
import org.luxons.sevenwonders.game.test.testTable

@RunWith(Theories::class)
class HandsTest {

    @Test(expected = IndexOutOfBoundsException::class)
    fun get_failsOnMissingPlayer() {
        val hands = createHands(4, 7)
        hands[5]
    }

    @Test
    fun get_retrievesCorrectCards() {
        val hand0 = sampleCards(0, 5)
        val hand1 = sampleCards(5, 10)
        val hands = Hands(listOf(hand0, hand1))
        assertEquals(hand0, hands[0])
        assertEquals(hand1, hands[1])
    }

    @Theory
    fun isEmpty_falseWhenAtLeast1_allSame(
        @FromDataPoints("nbPlayers") nbPlayers: Int,
        @FromDataPoints("nbCardsPerPlayer") nbCardsPerPlayer: Int
    ) {
        assumeTrue(nbCardsPerPlayer >= 1)
        val hands = createHands(nbPlayers, nbCardsPerPlayer)
        assertFalse(hands.isEmpty)
    }

    @Theory
    fun isEmpty_trueWhenAllEmpty(@FromDataPoints("nbPlayers") nbPlayers: Int) {
        val hands = createHands(nbPlayers, 0)
        assertTrue(hands.isEmpty)
    }

    @Theory
    fun maxOneCardRemains_falseWhenAtLeast2_allSame(
        @FromDataPoints("nbPlayers") nbPlayers: Int,
        @FromDataPoints("nbCardsPerPlayer") nbCardsPerPlayer: Int
    ) {
        assumeTrue(nbCardsPerPlayer >= 2)
        val hands = createHands(nbPlayers, nbCardsPerPlayer)
        assertFalse(hands.maxOneCardRemains())
    }

    @Theory
    fun maxOneCardRemains_trueWhenAtMost1_allSame(
        @FromDataPoints("nbPlayers") nbPlayers: Int,
        @FromDataPoints("nbCardsPerPlayer") nbCardsPerPlayer: Int
    ) {
        assumeTrue(nbCardsPerPlayer <= 1)
        val hands = createHands(nbPlayers, nbCardsPerPlayer)
        assertTrue(hands.maxOneCardRemains())
    }

    @Theory
    fun maxOneCardRemains_trueWhenAtMost1_someZero(@FromDataPoints("nbPlayers") nbPlayers: Int) {
        val hands = createHands(nbPlayers, 1)
        hands[0].drop(1)
        assertTrue(hands.maxOneCardRemains())
    }

    @Test
    fun rotate_movesOfCorrectOffset_right() {
        val hands = createHands(3, 7)
        val rotated = hands.rotate(HandRotationDirection.RIGHT)
        assertEquals(rotated[1], hands[0])
        assertEquals(rotated[2], hands[1])
        assertEquals(rotated[0], hands[2])
    }

    @Test
    fun rotate_movesOfCorrectOffset_left() {
        val hands = createHands(3, 7)
        val rotated = hands.rotate(HandRotationDirection.LEFT)
        assertEquals(rotated[2], hands[0])
        assertEquals(rotated[0], hands[1])
        assertEquals(rotated[1], hands[2])
    }

    @Test
    fun createHand_containsAllCards() {
        val hand0 = sampleCards(0, 5)
        val hand1 = sampleCards(5, 10)
        val hands = Hands(listOf(hand0, hand1))

        val table = testTable(2)
        val hand = hands.createHand(SimplePlayer(0, table))

        for (handCard in hand) {
            assertTrue(hand0.contains(handCard.card))
        }
    }

    companion object {

        @JvmStatic
        @DataPoints("nbCardsPerPlayer")
        fun nbCardsPerPlayer(): IntArray {
            return intArrayOf(0, 1, 2, 3, 4, 5, 6, 7)
        }

        @JvmStatic
        @DataPoints("nbPlayers")
        fun nbPlayers(): IntArray {
            return intArrayOf(3, 4, 5, 6, 7)
        }

        private fun createHands(nbPlayers: Int, nbCardsPerPlayer: Int): Hands {
            return sampleCards(0, nbCardsPerPlayer * nbPlayers).deal(nbPlayers)
        }
    }
}
